### PR TITLE
Abstract Raptor file system to use HDFS

### PR DESCRIPTION
**Pull request by highker**
```
== NO RELEASE NOTE ==
```

### CODE REVIEW

File: presto-raptor/src/main/java/com/facebook/presto/raptor/storage/StorageService.java [hunk 1]
```diff
@@ -23,13 +24,13 @@
...
+    void createParents(Path file);
 
-    File getStorageFile(UUID shardUuid);
+    Path getStorageFile(UUID shardUuid);
 
-    File getStagingFile(UUID shardUuid);
+    Path getStagingFile(UUID shardUuid);
 
-    File getQuarantineFile(UUID shardUuid);
+    Path getQuarantineFile(UUID shardUuid);
```
Comment (tdcmeehan) [inline presto-raptor/src/main/java/com/facebook/presto/raptor/storage/StorageService.java]: Could this all be framed instead in terms of URI, and we switch off the URI scheme to determine the FileSystem to use?

Comment (highker) [reply presto-raptor/src/main/java/com/facebook/presto/raptor/storage/StorageService.java]: I though about that. The reason to have `Path` is to work with `FileSystem` class (which all other Presto file systems like S3, GCS inherit from). That helps with `FileSystem` interfaces including `create`, `open`, `exists`, `delete`, `rename`, etc without converting from/to URI. Also, `Path` can support `getFileSystem` from a config as well. After checking how `HdfsEnvironment` class is used in our codebase, I decided to use `Path`/`FileSystem` combination. What do you think?

File: presto-raptor/src/main/java/com/facebook/presto/raptor/storage/ShardEjector.java [hunk 2]
```diff
@@ -112,6 +119,9 @@ public ShardEjector(
         this.interval = requireNonNull(interval, "interval is null");
         this.backupStore = requireNonNull(backupStore, "backupStore is null");
         this.executor = newScheduledThreadPool(1, daemonThreadsNamed("shard-ejector-" + connectorId));
+        this.localFileSystem = tryGetLocalFileSystem(requireNonNull(environment, "requireNonNull is null"));
```
Comment (tdcmeehan) [inline presto-raptor/src/main/java/com/facebook/presto/raptor/storage/ShardEjector.java]: Similar typo here

File: presto-raptor/src/main/java/com/facebook/presto/raptor/storage/ShardRecoveryManager.java [hunk 3]
```diff
@@ -79,6 +83,7 @@
 
     private final StorageService storageService;
     private final Optional<BackupStore> backupStore;
+    private final Optional<RawLocalFileSystem> localFileSystem;
```
Comment (tdcmeehan) [inline presto-raptor/src/main/java/com/facebook/presto/raptor/storage/ShardRecoveryManager.java]: Do we need to know that this is a local FS?

File: presto-raptor/src/main/java/com/facebook/presto/raptor/storage/OrcStorageManager.java [hunk 4]
```diff
@@ -709,7 +715,14 @@ private void createWriterIfNecessary()
                 File stagingFile = storageService.getStagingFile(shardUuid);
                 storageService.createParents(stagingFile);
                 stagingFiles.add(stagingFile);
-                writer = new OrcFileWriter(columnIds, columnTypes, stagingFile, orcOptimizedWriterStage.equals(ENABLED_AND_VALIDATED), stats, typeManager, compression);
+                OrcDataSink sink;
+                try (ThreadContextClassLoader ignored = new ThreadContextClassLoader(FileSystem.class.getClassLoader())) {
```
Comment (tdcmeehan) [inline presto-raptor/src/main/java/com/facebook/presto/raptor/storage/OrcStorageManager.java]: Why do we need to do this?

File: presto-raptor/src/main/java/com/facebook/presto/raptor/metadata/ShardCleaner.java [hunk 5]
```diff
@@ -83,6 +88,7 @@
     private final Duration backupCleanTime;
     private final ScheduledExecutorService scheduler;
     private final ExecutorService backupExecutor;
+    private final Optional<RawLocalFileSystem> localFileSystem;
```
[the remark this answers is not in the document]
Comment (highker) [reply presto-raptor/src/main/java/com/facebook/presto/raptor/metadata/ShardCleaner.java]: That is a good point. I was thinking making it explicit (like some other classes `ShardEjector`, `BackupManager`, etc) to indicate these classes only work with local FS. It does not quite make sense for them to serve disagg purpose. Some classes like `OrcStorageManager` will have the interface `FileSystem` given they will deal with various FS's.

**Comment by highker** [comment]
Thanks @tdcmeehan for the review. Comments addressed.

File: presto-raptor/src/main/java/com/facebook/presto/raptor/storage/StorageService.java (hunk 1 above)
Comment (tdcmeehan) [reply presto-raptor/src/main/java/com/facebook/presto/raptor/storage/StorageService.java]: Internally, Path contains the URI and will internally switch off the scheme.  So my suggestions was not that useful.

File: presto-raptor/src/main/java/com/facebook/presto/raptor/metadata/ShardCleaner.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (jessesleeping) [reply presto-raptor/src/main/java/com/facebook/presto/raptor/metadata/ShardCleaner.java]: I think it make sense to confine it to local file system here in the `ShardCleaner`. 

I think for `ShardEjector`, `BackupManager` and `ShardRecoveryManager`, they only make sense with local file system. We still need some shard cleaner after we move to disagg, but the current implementation is very specific to local file system. 

Also, I am feeling that `ShardEjector`, `BackupManager`, `ShardRecoveryManager` and `ShardCleaner` should all be hidden under a storage layer interface as disagg and local file system have completely different requirement on these components.

File: presto-raptor/src/main/java/com/facebook/presto/raptor/storage/OrcStorageManager.java [hunk 6]
```diff
@@ -346,12 +359,13 @@ private void writeShard(UUID shardUuid)
...
-        File stagingFile = storageService.getStagingFile(shardUuid);
-        File storageFile = storageService.getStorageFile(shardUuid);
+        File stagingFile = localFileSystem.pathToFile(storageService.getStagingFile(shardUuid));
+        File storageFile = localFileSystem.pathToFile(storageService.getStorageFile(shardUuid));
 
-        storageService.createParents(storageFile);
+        storageService.createParents(new Path(storageFile.toURI()));
 
         try {
+            // Do not use FileSystem::moveXXX which does a copy and delete
```
[the remark this answers is not in the document]
Comment (highker) [reply presto-raptor/src/main/java/com/facebook/presto/raptor/storage/OrcStorageManager.java]: This has already been updated in a later commit.

File: presto-raptor/src/main/java/com/facebook/presto/raptor/filesystem/RaptorLocalFileSystem.java [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+        extends RawLocalFileSystem
+{
+    public RaptorLocalFileSystem(Configuration configuration)
+            throws IOException
+    {
+        initialize(getUri(), configuration);
+    }
+
+    @Override
+    public boolean exists(Path file)
```
Comment (jessesleeping) [inline presto-raptor/src/main/java/com/facebook/presto/raptor/filesystem/RaptorLocalFileSystem.java]: Can we add a comment explaining why we specifically want to overwrite these 2 methods (`exists()` and `rename()`)?

File: presto-raptor/src/main/java/com/facebook/presto/raptor/storage/ShardEjector.java [hunk 8]
```diff
@@ -242,8 +253,8 @@ void process()
             shardManager.replaceShardAssignment(shard.getTableId(), shardUuid, target, false);
 
             // delete local file
-            File file = storageService.getStorageFile(shardUuid);
-            if (file.exists() && !file.delete()) {
+            Path file = storageService.getStorageFile(shardUuid);
```
Comment (jessesleeping) [inline presto-raptor/src/main/java/com/facebook/presto/raptor/storage/ShardEjector.java]: Why don't we just convert `Path` to `File` and use OS file system call like what we did in `BackupManager`?

Comment (jessesleeping) [reply presto-raptor/src/main/java/com/facebook/presto/raptor/storage/ShardEjector.java]: I think in more generic classes like `OrcPageRewriter` and `OrcStorageManager` we can stick to using `Path`. But for local flash specific components like `BackupManager`, `ShardEjector`, `ShardRecoveryManager` and `ShardCleaner` maybe we can always convert `Path` to `File` when invoking file system related call.

File: presto-raptor/src/main/java/com/facebook/presto/raptor/metadata/ShardCleaner.java [hunk 9]
```diff
@@ -513,8 +526,9 @@ private static Timestamp maxTimestamp(Duration duration)
     }
 
     @SuppressWarnings("ResultOfMethodCallIgnored")
-    private static void deleteFile(File file)
+    private void deleteFile(Path file)
+            throws IOException
     {
-        file.delete();
+        localFileSystem.get().delete(file, false);
```
Comment (jessesleeping) [inline presto-raptor/src/main/java/com/facebook/presto/raptor/metadata/ShardCleaner.java]: Same comment as in `ShardEjector`.

File: presto-raptor/src/main/java/com/facebook/presto/raptor/storage/ShardEjector.java (hunk 8 above)
Comment (highker) [reply presto-raptor/src/main/java/com/facebook/presto/raptor/storage/ShardEjector.java]: Two reasons:
- To work on `File`, we need to call `RawLocalFileSystem::pathToFile` to convert `Path` to `File`. That requires an explicit file system anyway.
- Also, it provides better abstraction.

**Comment by highker** [comment]
Comments addressed; thanks for review

File: presto-raptor/src/main/java/com/facebook/presto/raptor/backup/BackupManager.java [hunk 10]
```diff
@@ -52,36 +52,31 @@
 public class BackupManager
 {
     private static final Logger log = Logger.get(BackupManager.class);
-    private static final Configuration CONFIGURATION = new Configuration();
 
     private final Optional<BackupStore> backupStore;
     private final StorageService storageService;
     private final ExecutorService executorService;
-    private final RawLocalFileSystem localFileSystem;  // TODO: inject FileSystem once HDFS environment is ready
+    private final Optional<RawLocalFileSystem> localFileSystem;
```
[the remark this answers is not in the document]
Comment (highker) [reply presto-raptor/src/main/java/com/facebook/presto/raptor/backup/BackupManager.java]: It is coupled with `Optional<BackupStore> backupStore`, where they need to be both present or empty. Of course, there is a corner case that backup does not have to be present even with local FS.

File: presto-raptor/src/main/java/com/facebook/presto/raptor/storage/ShardEjector.java (hunk 8 above)
Comment (jessesleeping) [reply presto-raptor/src/main/java/com/facebook/presto/raptor/storage/ShardEjector.java]: I am not very strong on this. Here is my thoughts:

For classes like `OrcPageRewriter` and `OrcStorageManager`, they will be used by both local and disagg storage. So they have to operate on hdfs `Path`, not `File`. But for local storage specific classes (`BackupManager`, `ShardEjector`, `ShardRecoveryManager` and `ShardCleaner`), they anyway need a local file system to operate on. Using `File` and calling OS file system call (i.e. `file.delete()`, `file.exists()`) makes sure their behavior is exactly the same as previous and introduces minimum changes.

**Pull request merged**